### PR TITLE
Implement finalizer

### PR DESCRIPTION
**Pull request by atanasdinov**
- Introduces a cleanup logic for deletion requests
- The SUC plans and Secret resources will be present on the cluster until the respective UpgradePlan is removed

### CODE REVIEW

File: internal/upgrade/os.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 
 const (
 	scriptName = "os-upgrade.sh"
+	SecretName = "os-upgrade-secret"
```
Comment (ipetrov117) [inline internal/upgrade/os.go:20]: Where do we use this outside of the `upgrade` package?

File: internal/controller/upgradeplan_controller.go [hunk 2]
```diff
@@ -80,13 +81,78 @@ func (r *UpgradePlanReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+		if !controllerutil.ContainsFinalizer(plan, lifecyclev1alpha1.UpgradePlanFinalizer) {
+			return ctrl.Result{}, nil
+		}
+
+		if err := r.reconcileDelete(ctx, plan); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		controllerutil.RemoveFinalizer(plan, lifecyclev1alpha1.UpgradePlanFinalizer)
+		return ctrl.Result{}, r.Update(ctx, plan)
```
Comment (ipetrov117) [inline internal/controller/upgradeplan_controller.go:94]: Currently for the use-case where a user attempts to delete an `UpgradePlan` where its underlying `SUC Plans` are still being applied we get the following:
1. `r.reconcileDelete()` -> SUC Plan still applying -> return nil
2. Attempt to delete finaliser -> Upgrade
3. Reconcile -> reconcileDelete -> If plan is still applying redo (1) and (2)

So some questions here:
1. Shouldn't the `UpgradePlan` finaliser be removed only after the underlying SUC Plans and OS secret are removed?
2. If we do a finaliser remove + `UpgradePlan` update while SUC Plans are still applying, what will this result in?
2.1. Will it result in a rerun of the reconciliation loop?
2.2. Will it delete the `UpgradePlan` without actually deleting the SUC Plans + OS secret?

**Review by ipetrov117 — COMMENTED** [review]
Great work, will be happy to approve after the underlying comments are discussed.

File: internal/upgrade/os.go (hunk 1 above)
Comment (atanasdinov) [reply internal/upgrade/os.go]: Leftover from the initial implementation. The idea there was that we'd be reusing the secret so it was fine to just issue a single delete command, however, we decided that we'll be using different secrets so I adjusted to list them as well. Removed, thanks!

**Review by ipetrov117 — APPROVED** [review]

**Pull request merged**